libp2p/protocols/kademlia/rpc/provider.nim

```diff
@@ -0,0 +1,63 @@
+# Nim-Libp2p
+# Copyright (c) 2023 Status Research & Development GmbH
+# Licensed under either of
+#  * Apache License, version 2.0, ([LICENSE-APACHE](LICENSE-APACHE))
+#  * MIT license ([LICENSE-MIT](LICENSE-MIT))
+# at your option.
+# This file may not be copied, modified, or distributed except according to
+# those terms.
+
+import std/sequtils
+import chronos, chronicles, results
+import ../../../[peerid, switch, multihash]
+import ../../protocol
+import ../[protobuf, types]
+import ./find
+
+proc dispatchAddProvider(
+    switch: Switch, peer: PeerId, key: Key
+) {.async: (raises: [CancelledError, DialFailedError, LPStreamError]).} =
+  let conn = await switch.dial(peer, KadCodec)
+  defer:
+    await conn.close()
+
+  let msg = Message(
+    msgType: MessageType.addProvider,
+    key: Opt.some(key),
```

Comment:
how can `key` be optional?

Reply:
Good question, I'll take a look and since this would touch all rpc methods, I'll implement this in another PR: https://github.com/vacp2p/nim-libp2p/issues/1756?issue=vacp2p%7Cnim-libp2p%7C1774